### ISSUE TITLE
-Wno-sign-compare revisited 

### ISSUE THREAD

**Issue by moskewcz** [BVLC/caffe#2216]
following up on PR#488, based on running:
git log -G'sign-compare' -- Makefile

it looks like the warning settings for caffe were, in chrono order:

1) -Wall
2) -Wall  -Werror -Wno-sign-compare
3) -Wall -Wno-sign-compare

i just started noticing sign-compare warning when including caffe headers in another project using 'straight' -Wall. there aren't too many (11 by my count) sign-compare warnings in caffe *headers* (there are 101 more in .cpp files). if we can agree on a pattern to fix/supress them, what are the chances that a PR to that effect might get merged? my understanding that, in these cases the compiler is casting the int32_t to uint32_t and then doing uint32_t comparison, so the 'normal' fix is to explicity cast the signed value to unsigned.

moskewcz@orihime:~/git_work/caffe$ less make.log  | grep warning | grep .hpp | sort  -u
./include/caffe/blob.hpp:172:7: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/blob.hpp:57:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:169:67: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:176:5: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:275:43: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:282:5: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:388:25: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:412:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:423:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/layer.hpp:462:16: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
./include/caffe/loss_layers.hpp:76:18: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
moskewcz@orihime:~/git_work/caffe$

**Comment by moskewcz** [comment]
looking a one actual case in blob.hpp, i'd say the sanest fix is to use size_t instead of int, which is at least (arguably) closer to the return type of vector<int>.size() (technically vector<int>::size_type). 

it would seem you'd have to be pretty hard-core against unsigneds to have a problem with this change, given that you can't change the return type of size() and using an unsigned here seems totally correct, sensible, and preferable to using an int ... but as usual in c++, it seems you can't every really win or please everyone.

     ostringstream stream;
-    for (int i = 0; i < shape_.size(); ++i) {
+    for (size_t i = 0; i < shape_.size(); ++i) {
       stream << shape_[i] << " ";

**Comment by longjon** [comment]
The omission of `-Werror` is, I hope, temporary.

The use of signed `int` and `-Wno-sign-compare` is intentional, due to http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types, though I can see the argument the other way in this case. Maybe we should reconsider the use of iterators if/when we switch to C++11.

**Comment by moskewcz** [comment]
oh, sure, i realize you're refing the famous google style guide. it's awfully restrictive overall, and i personally don't find its reasoning particualy compelling or well thought out in general (although i generally agree with many particular points).

in this particular case, the advice given seems ... questionable at best. it even admits there is a tradeoff but simply states that it swings a certain way. the one example given seems, if anything, to indicate to me some logical bug that would be *hidden* by using a signed comparison -- i.e. getting an infinite loop there would be diagnostic of a missing/incorrect assumption. but i digress; even if we were to decide to follow the rule without debate, i'm not sure it's really suggesting blind use of signed types everwhere in all code that interacts with existing code (i.e. STL in this case). in particular, it calls out the *danger* of mixing signed with unsigned, which is exactly what you're doing here, and you had to disable a solid compiler warning to get a clean compile. i think since, in this case, you are *forced* to use an unsigned type as the return from size(), you should match types. you could do this either by casting the size() to int, or by using an unsigned as the iterator type. arguably i'd say a strict reading of the rule would require a wrapper around size that checked for overflow before casting to int. then, if the container can actually hold more items than can fit in an int, it's going to be hard to iterate over it with ints -- would you then propose using int64_t or the like, and checking that the size type is only 32 bit? again, that seems pretty questionable. i'd note the guide also calls out this general principle specifically:
"You may diverge from the rules when dealing with code that does not conform to this style guide."
and:
"Use common sense and BE CONSISTENT."

the guide also states that when using ints that logically must be >=0, this should be documented with asserts(). i didn't comb the caffe code, but i'm guessing there may be a ton of such asserts() that are missing. i think if the choice is between code using all ints *without* the needed asserts() for non-negative things and acutally using unsigned vars where it makes sense, i can make a strong agument that using unsigned is superior. again, the rule is not about blinding using ints everywhere, but using ints in a certain set of cases with a certain amount of rigor and extra checking.

i'd also note the advice on the same page is to always use specific bit-width types, which is certainly not being done here -- and again i'm not sure it makes sense in general or as a hard rule. i'm all for using specific widths were possible, but i think matching types when you can't control them (i.e. size_t, templated/nested typedefs) trumps that rule.

a relevant SO Q:
http://stackoverflow.com/questions/4377948/how-to-ignore-comparison-between-signed-and-unsigned-integer-expressions

**Comment by longjon** [comment]
Yes, as I said, I can see the argument in this case. However, the argument against `unsigned` types isn't negated; doing arithmetic on unsigned loop counters would still be subject to those pitfalls.

The guide does not advise always using specific bit-width types. In fact, it states

    We use int very often, for integers we know are not going to be too big, e.g., loop counters. Use plain old int for such things.

So overall my feeling, at least for this very common idiom, is that we should stick to the plain `int` version, and punt on the issue of the comparison mismatch until C++11; iterators do seem like the most correct approach. But I'd welcome varying opinions from other Caffe devs.

**Comment by moskewcz** [comment]
ah, yep, i misread the bit-width section.

**Comment by moskewcz** [comment]
i also wonder if the intent is that caffe can be used as a library in C++ or not; and if so, which headers constitute the public interface.

it's one thing to follow the style guide of one's choice internally, but it seems questionable to publish a general-audience (i.e. not google internal) library where the interface headers can't be compiled warning free with -Wall. the google style guide seems notable silent on compiler/warning options, and it's still not clear to me that they really advocate disabiling the warning in question (although i guess i see some evidence that that is indeed the case).

i can't think of another library offhand, C++ or C, where the headers don't compile warning-free with -Wall -- but, again, i dunno if caffe falls into that sort of category or not.

**Comment by longjon** [comment]
Yep, Caffe-as-library is due for additional consideration at some point. Currently it's quite possible to use Caffe that way, but we don't guarantee, e.g., that basic interfaces will not change. I agree that in that case having warnings in the header files is awkward; maybe the inline functions causing them don't really need to be inline.

**Comment by moskewcz** [comment]
really, i'm not too picky about what change is made as long as it silences the warning; in addition to de-inlining the code, here are 3 more options that don't involve using an unsigned if you so insist:

option 1) preserve the existing (bad IMO) semantics, but make the cast explicit to avoid the warning:
````
     ostringstream stream;
-    for (int i = 0; i < shape_.size(); ++i) {
+    for (int_t i = 0; (size_t)i < shape_.size(); ++i) {
     stream << shape_[i] << " ";
````

option 2: cast the size() to int: this is 'less unsigned', but isn't quite the same as the original semantics. it's arguably more correct though.
````
     ostringstream stream;
-    for (int i = 0; i < shape_.size(); ++i) {
+    for (int_t i = 0; i < (int)shape_.size(); ++i) {
     stream << shape_[i] << " ";
````
option 3: use a regular pre-c++89 iterator:
````
     ostringstream stream;
-    for (int i = 0; i < shape_.size(); ++i) {
+    for (vector<int>::const_iterator i = shape_.begin(); i != shape_.end(); ++i) {
     stream << (*i) << " ";
````

and, bear in mind the warning-silenced type of for loop test you seem to be defending seems to exhibit exactly the sort of unsigned-ness bug potential/confusion that the 'avoiding unsigned' rule is supposed to prevent, since it does in fact mix signed and unsigned types:
````
#include<vector>
#include<cstdio>
int main( void ) {
  unsigned iters = 0; std::vector<int> foo( 3 );
  for (int i = -2; i < (int)foo.size(); ++i) { ++iters; }
  printf("iters=%u\n",iters); return 0; 
}
moskewcz@orihime:~/scratch$ g++ -Wall int_loop.cc  # note: no warnings
moskewcz@orihime:~/scratch$ ./a.out 
iters=5

#include<vector>
#include<cstdio>
int main( void ) {
  unsigned iters = 0; std::vector<int> foo( 3 );
  for (int i = -2; i < foo.size(); ++i) { ++iters; }
  printf("iters=%u\n",iters); return 0; 
}
moskewcz@orihime:~/scratch$ g++ -Wall int_loop.cc 
int_loop.cc: In function ‘int main()’:
int_loop.cc:5:33: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
moskewcz@orihime:~/scratch$ ./a.out 
iters=0 # probably not the 'expected'/correct result, depends on int wraparound, seems confusing to me

````

**Comment by bhack** [comment]
We already use caffe as library but for this scope headers need to be reorganized. I think that was not thought what is the public api  we want to expose for third party c++ apps.
